README.md

```diff
@@ -1,2 +1,78 @@
-# example-application
-Example out-of-tree application that is also a module
+# Zephyr Example Application
```

Comment:
>  (From @carlescufi) Do not expect that this repository will ever build .rst files (or be built as part of Zephyr documentation). Leave all documentation in Markdown.

Agree for the README.

>  (From @galak) Make sure it is organized to support multiple example applications in the same repository.  This will change your code layout.

Perhaps we should start by removing the docs entirely before trying to address this goal in a later PR.

> It is not well-organized for extending existing repos to also be Zephyr modules. 

My impression of the dev-review discussion was that this is a non-goal for this initial effort.

Reply:
> 3. It is not organized to grow to completeness.
>  i. Any driver, subsystem, or application example code needs to be covered by test cases. This will affect the settings in zephyr/module.yml and change your code layout. (How can you unit test anything beneath app/src/?)

Where in Zephyr are we unit testing anything in `<app>/src`?

As far as I know, apps are tested by running them in twister and seeing what happens, not unit tested.

> ii. A driver example needs to include a custom interface header declaring and documenting the driver API.

Are you suggesting this sample should add an entirely new driver subsystem, rather than just implementing an existing one?

If so, I strongly disagree. If not, what do you mean?

> iii. Since this includes generated documentation, it needs a local sample that is integrated into that documentation. This will affect the settings in zephyr/module.yml and expand your code layout.

I am really starting to think that solving the docs question by not having any to start with is the way to go...

> iv. This needs to include files necessary for automated code coverage measurements and reporting, especially since the Zephyr CI system discovers tests and samples through module.yml. This will add testcase.yml and sample.yml to your code tree.

Just sample.yaml for the app, right? I don't see any tests/ directory in this PR.

> v. When this repository supports multiple applications (conf.prj) it will become obvious that FW "version" really tracks with the repository shaw, not a sub-tree of the application. The whole "version" implementation really needs to be more thoroughly designed and explained.

How about just removing it from the PR for now then?

Reply:
Some answers to the questions:

> In general I have four issues with this PR:
> 
> 1. This does not meet the complaints noted about the first PR.
>    
>    1. Keep it **SIMPLE** - The first PR also started out with 41 files and was not acceptable even when trimmed down to 10.

Simplicity should not be strictly measured using the number of files. It probably has some correlation, but a big number of code files may be considered simple, e.g. if they offer a clean API, result in simple usage, etc.

>    2. (From @carlescufi) Do not expect that this repository will ever build .rst files (or be built as part of Zephyr documentation). Leave all documentation in Markdown.

I don't get this point. The example just shows how to:

- Use Sphinx + Breathe
- Reference Zephyr pages using intersphinx

> 2. It does not meet the scalability goals noted during the dev review of the first PR.
>    
>    1. (From @galak) Make sure it is organized to support multiple example applications in the same repository.  This will change your code layout.

If the "example-application" needs to support multiple applications it should likely be named "example-application*s*". Regarding scalability, there can be other folders with their own set of CMakeLists.txt, prj.conf, etc. So I don't see a problem in that aspect. Still, I don't think it should be the goal of this example, as it is a more advanced usecase which also has other solutions (e.g. a repo for each app with a module for shared codebase).

> 3. It is not organized to grow to completeness.
>    
>    1. Any driver, subsystem, or application example code needs to be covered by test cases. This will affect the settings in `zephyr/module.yml` and change your code layout.  (How can you unit test anything beneath `app/src/`?)

If we add tests in the example-application they should be added for demonstration purposes, e.g. how do I test my app? Not to test all demonstrated features as if it was a real codebase. example-application can point to real application for these sort of things.

>    2. A driver example needs to include a custom interface header declaring and documenting the driver API.

I think this should be out of the scope of a basic example-application

Reply:
> Simplicity should not be strictly measured using the number of files. It probably has some correlation, but a big number of code files may be considered simple, e.g. if they offer a clean API, result in simple usage, etc.

Oh, I completely understand and agree.  Keeping it SIMPLE is the perpetual feedback @carlescufi gave to PR1, so it should be used to evaluate this PR also.  Eventually I learned he wanted it small enough to be completely understandable and reviewable in about 5 minutes.  This led to continually cutting PR1 down to a near trivial amount of code and creating several RFCs to grow the additional functionality.

I actually prefer the **Do The Simplest Thing _That Can Possibly Work_ (DTSTTCPW)** guidance, where "Work" includes automated verification of required functionality.  In this case required functionality is exercising all the control points in `module.yml` and that all the config/build/doc generation tools that are correctly and completely following each.

> I don't get this point.

I don't get it either, but it was provided as feedback about PR1.  I and others have found the documentation system to be a very valuable aspect of Zephyr.  Since requirements tracing will depend on Doxygen tags it seems that document generation is as central to the Zephyr solution as `twister`.  Perhaps the guidance was because configuring and maintaining the documentation tools is non-trivial (as I am learning)?

> If the "example-application" needs to support multiple applications it should likely be named "example-applications".

I spent a LOT of time trying to negotiate a naming pattern for "oot" repositories hosted in the zephyrproject-rtos organization, and other participants just weren't ready to accept the inevitability of it.  So the direction was set to put it all in to one repository.  This meant that the repository needs to hold a range of content that can **easily** be stripped down to just what is desired (e.g. one custom board).  I ended up being comfortable with the name "example-application" once I recognized the "application" portion referred to the role of the repository in the different topologies (star, forest, ...) rather than the type, number, or scope of content within it.

> Regarding scalability, there can be other folders with their own set of CMakeLists.txt, prj.conf, etc. 

Yes, there **can** be other folders parallel to `app/`, but making `app/` a tree (similar to `drivers/', 'subys/', 'samples/') naturally is paralleled under `tests/`.  This is one area where including tests in the PR now will avoid restructuring it later, which is of great benefit to all.  I strongly prefer not starting people off with a structure they will (very) quickly outgrow, or with a structure that will have to change as we add more breadth of extensibility.

> Still, I don't think it should be the goal of this example, 

You now touched on the key issue getting in the way of accepting PR1: What is the goal of this example? What is SIMPLE?  Who is the audience?  There never has been clear alignment on this, leading to frustrations on all sides.

From the PR2 README.md:

> The main purpose of this repository is to serve as a reference on how to structure Zephyr based applications.

No quite.  Before I ever approached Carles about creating this repository I understood the current and future extensiblity enabled by `module.yml`, wanted to depend on it for my own scalable solution, and could not find any evidence that the extensibility was being fully regressively verified.  Recognizing that verification required using a repository outside the default Zephyr manifest, hearing in the Process Working Group about the desire/benefit of an application manifest, and expecting the Zephyr Project wanted to minimize the number of repositories to be maintained I proposed we make one repository serving as **primarily** an extensibility test case (needed for verifying Zephyr) and **secondarily** as a reference for an OOT module.  I have made sure that usage as a test case was never excluded from the scope.  (BTW, the _current_ config/build/doc generation tool set do not consistently or completely follow this mechanism.)

The quoted purpose does not indicate anything about simple vs more advanced used cases, and implies this is the recommended way.  Requiring more repositories for each application is inconsistent with the policy used on Zephyr itself, where everything goes into one repository and (in general) it is set up for scalability.

Adjusting this PR to fulfill the testing needs (requirements) will impact the scope statement and configuration of this repository in two ways: organization and breadth of features. In order to verify that a `module.yml` setting is being used there must be two test points.  Zephyr itself provides the top-level paths, so this repository **should not** have any of the files referenced in `module.yml` at the top level.  The breadth will drive at least `tests` and `sampto maximizeles` into the directory tree.

> as it is a more advanced use case which also has other solutions (e.g. a repo for each app with a module for shared codebase).

That solution is always an option, but not also setting it up for a tree of applications in this single example sends the message that Zephyr won't support scalability in the application, thus "removing" user choices.  This was a criticism voiced against PR1 at one point, so must be considered here too.

> If we add tests in the example-application they should be added for demonstration purposes, e.g. how do I test my app? Not to test all demonstrated features as if it was a real codebase. example-application can point to real application for these sort of things.

Ah, you missed the impetus and foundational purpose for this repository.  It **is** a real codebase with richly connected trivial application functionality used needed to test all supported features that are extended through `module.yml`. This was always proposed as a boilerplate/template including correct and extensive error handling logic.

> My impression of the dev-review discussion was that this is a non-goal for this initial effort.

If it was to be a non-goal for the initial effort then there was no benefit in requesting it then. Given the comment was around repository scope rather than an individual PR I understood it must be set up (organized) to support it with natural growth.  

> Where in Zephyr are we unit testing anything in <app>/src?
> 
> As far as I know, apps are tested by running them in twister and seeing what happens, not unit tested.

We should consider anything run by twister as an app, since it generates a runnable executable.  Twister runs and can evaluate the console output of any "app".  When the "app" contains ztest test cases, the test runner fulfills the role of the "app", whose purpose is to unit-test bound libraries.  Since Zephyr is adopting MISRA guidelines that there is no (unjustified) dead or deactivated code, all logic not traced back to requirements and not verified (or justified by human inspection per release) will need to be removed.  A natural and automatable solution is to minimize the logic in places that cannot be covered by automated unit tests.  So, logic goes in subsystems, and `main()` becomes empty.

> Are you suggesting this sample should add an entirely new driver subsystem, rather than just implementing an existing one?

Almost.  A new one is needed to verify Zephyr build system functionality.  Implementing an existing one is needed to verify an API spec'd in Zephyr can be implemented OOT. So ... both!

> Just sample.yaml for the app, right? I don't see any tests/ directory in this PR.

Tests are explicitly in scope for this repository. See the last bullet in the initial entry of https://github.com/zephyrproject-rtos/zephyr/issues/28438.

> How about just removing it from the PR for now then?

I'm good with removing version reporting from this PR.  We will need to design and implement a solution at some point.  In PR1 I followed the same version pattern and mechanisms as used in Zephyr.

> I am really starting to think that solving the docs question by not having any to start with is the way to go...

I applaud the goal, effort, and the detailed comments.  I think we should hold off on the implementation until we get alignment on what use cases the Zephyr Project needs to support and design a structure and mechanism that meets those requirements.

Reply:
>The first PR also started out with 41 files and was not acceptable even when trimmed down to 10.

But that was, IIRC, without a driver or board, so file count is not comparable here.

>(From @galak) Make sure it is organized to support multiple example applications in the same repository. This will change your code layout.

I actually went back and checked with him, and he and I agree that the only thing we don't want is multiple `example-application-n` repositories. **If** demand exists for multiple applications to be supported, we can make that change at that point.

>How can you unit test anything beneath app/src/

Unit testing was never part of the [original description](https://github.com/zephyrproject-rtos/zephyr/issues/28438). A `tests/` folder can be added later.

>A driver example needs to include a custom interface header declaring and documenting the driver API.

Only if you are writing a new driver subsystem, as @mbolivar-nordic mentioned. Let us not go there yet.

>Since this includes generated documentation, it needs a local sample that is integrated into that documentation. This will affect the settings in zephyr/module.yml and expand your code layout.

I honestly do not understand what you mean. Documentation has nothing to do with `zephyr/module.yml`, and the doc for the app is in the corresponding `.rst` file.

>This needs to include files necessary for automated code coverage measurements and reporting,

No it does not, we never intended to cover that aspect of software development.

>When this repository supports multiple applications (conf.prj) it will become obvious that FW "version" really tracks with the repository shaw, not a sub-tree of the application. 

And when that happens, if it ever happens, we will deal with it.

>It is not well-organized for extending existing repos to also be Zephyr modules

This was never a goal, and it is not a goal now. The goal of this repository, and I quote from the [original description](https://github.com/zephyrproject-rtos/zephyr/issues/28438):

"Create a new repository in the https://github.com/zephyrproject-rtos organizationt that can serve both as reference and as a starting point for users wanting to create their own manifest+module repo"

This Pull Requests satisfies that premise, and it covers all the aspects mentioned in it except for the library and the tests.

Reply:
> > It is not well-organized for extending existing repos to also be Zephyr modules
> 
> This was never a goal, and it is not a goal now. 

From the email thread that kicked off that meeting:

> From: Cufi, Carles Carles.Cufi@nordicsemi.no
> Sent: Tuesday, August 25, 2020 8:00 AM
> To: Gregory A Shue gregory.shue@legrand.us; Rasmussen, Torsten Torsten.Rasmussen@nordicsemi.no; Glaropoulos, Ioannis Ioannis.Glaropoulos@nordicsemi.no; Bolivar, Marti Marti.Bolivar@nordicsemi.no
> Subject: RE: Repository structure and modules
> 
> Sure, I've added it to the meeting.
> 
> -----Original Message-----
> From: Gregory A Shue gregory.shue@legrand.us
> Sent: 20 August 2020 17:31
> To: Cufi, Carles Carles.Cufi@nordicsemi.no; Rasmussen, Torsten
> Torsten.Rasmussen@nordicsemi.no; Glaropoulos, Ioannis
> Ioannis.Glaropoulos@nordicsemi.no; Bolivar, Marti
> Marti.Bolivar@nordicsemi.no
> Subject: RE: Repository structure and modules
> 
> For this meeting, can we use
> https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgith
> ub.com%2Fproject-&data=02%7C01%7Cgregory.shue%40legrand.us%7C11fee
> e6cb44c4eef9b0e08d8490789ec%7C199686b5bef4496087867a6b1888fee3%7C1%7C0
> %7C637339644012170550&sdata=72EZnhce9N3JFBwN6UnhgOU9riSmXqt%2F4xe%
> 2BuiZeywk%3D&reserved=0 chip/connectedhomeip.git as an example of
> a repository needing to be extended into a Zephyr module?
> 
> Thanks,
> 
> Gregory Shue

I would not have agreed to the description if it excluded this case.  Go back and carefully read what the description text actually says.  This was always in scope and under consideration.